R/tspa_plot_revised.R

```diff
@@ -0,0 +1,123 @@
+tspa_plot <- function(tspa_fit,
+                      ask = FALSE,
+                      title = NULL,
+                      label_x = NULL,
+                      label_y = NULL,
+                      abbreviation = TRUE,
+                      fscore_type = NULL,
+                      ...) {
+
+  fit_pars <- lavaan::parameterestimates(tspa_fit) # parameter estimates
+  regression_fit <- fit_pars[which(fit_pars$op == "~"),]
+  latent_dv <- unique(c(unname(t(regression_fit["lhs"])))) # Extract DV
+  latent_iv <- unique(c(unname(t(regression_fit["rhs"])))) # Extract IV
+  fscores_func <- lavaan::lavInspect(tspa_fit, what = "data") # factor scores from `get_fs`
+  fscores_est <- lavaan::lavPredict(tspa_fit) # factor scores from estimation using `lavPredict`
+
+    # Type of scatterplot
+    if (is.null(fscore_type)) {
+      fscores <- fscores_est
+    } else {
+      fscores <- fscore_func
+    }
+
+    # Helper function for scatter plot
+    # Comment: put it outside the main function
+    plot_scatter <- function (fscores_df, iv, dv,
+                              ab_slope, ab_intercept, g_num = NULL, g_name = NULL, ...) {
+
+      if (!is.null(g_name)) {
+        iv_data <- as.data.frame(unname(fscores_df))[ ,iv]
+        dv_data <- as.data.frame(unname(fscores_df))[ ,dv]
+      } else {
+        iv_data <- fscores_df[ ,iv]
+        dv_data <- fscores_df[ ,dv]
+      }
+
+      plot(iv_data,
+           dv_data,
+           ylab = ifelse(is.null(label_y), paste0("fs_", dv),
```

Comment:
From the documentation of `ifelse`:

> Further note that if(test) yes else no is much more efficient and often much preferable to ifelse(test, yes, no) whenever test is a simple true/false result, i.e., when length(test) == 1.

And it doesn't seem the variable `i` is defined.

Reply:
Hi Mark, I revised this part as well. Thanks for the suggestion.